go/libkb/per_user_key.go

```diff
@@ -0,0 +1,598 @@
+package libkb
+
+import (
+	"crypto/rand"
+	"crypto/subtle"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"sync"
+
+	keybase1 "github.com/keybase/client/go/protocol/keybase1"
+	"github.com/keybase/go-codec/codec"
+	"golang.org/x/crypto/nacl/secretbox"
+	context "golang.org/x/net/context"
+)
+
+const PerUserKeySeedSize = 32
+
+// A secretbox containg a seed encrypted for its successor generation
+type PerUserKeyPrev string
+
+type PerUserKeySeed [PerUserKeySeedSize]byte
+
+func (s *PerUserKeySeed) DeriveSigningKey() (*NaclSigningKeyPair, error) {
+	derived, err := DeriveNonReversed(*s, DeriveReasonPUKSigning)
+	if err != nil {
+		return nil, err
+	}
+	res, err := MakeNaclSigningKeyPairFromSecret(derived)
+	return &res, err
+}
+
+func (s *PerUserKeySeed) DeriveEncryptionKey() (*NaclDHKeyPair, error) {
```

Comment:
Might make sense to call this a DH key or something, because there might also be times we derive symmetric encryption keys? (Besides the prev key...key :) ...which has its own method below.)